File: openapi-to-rust-bindings/src/openapi_to_rust_bindings/parser.py
"""Read openapi-to-rust generated sources as rust-sdk-generator Bindings."""

from __future__ import annotations

from pathlib import Path
import re

from tree_sitter import Language, Node, Parser
import tree_sitter_rust

from .model import Bindings


class ParseError(ValueError):
    """Generated Rust cannot be normalized into the supported Bindings contract."""


def _text(source: bytes, node: Node | None) -> str:
    if node is None:
        raise ParseError("missing Rust AST node")
    return source[node.start_byte : node.end_byte].decode()


def _children(node: Node | None, kind: str) -> list[Node]:
    return [child for child in node.named_children if child.type == kind] if node else []


def _serde_rename(attributes: list[str]) -> str | None:
    matches = []
    for attribute in attributes:
        match = re.fullmatch(r'\#\[serde\(rename\s*=\s*"([^"]+)"\)\]', attribute)
        if match:
            matches.append(match.group(1))
    if len(matches) > 1:
        raise ParseError("multiple serde rename attributes on enum variant")
    return matches[0] if matches else None


def _split_generic(value: str) -> tuple[str, list[str]] | None:
    start = value.find("<")
    if start < 0 or not value.endswith(">"):
        return None
    constructor = value[:start].strip()
    body = value[start + 1 : -1]
    depth = 0
    part_start = 0
    parts: list[str] = []
    for index, character in enumerate(body):
        if character in "<([{":
            depth += 1
        elif character in ">)]}":
            depth -= 1
        elif character == "," and depth == 0:
            parts.append(body[part_start:index].strip())
            part_start = index + 1
    parts.append(body[part_start:].strip())
    return constructor, parts


def _stream_binding(success_type: str) -> dict[str, str] | None:
    outer = _split_generic(success_type)
    if outer is None:
        return None
    constructor, arguments = outer
    if constructor != "futures_util::stream::BoxStream" or len(arguments) != 2:
        return None
    event = _split_generic(arguments[1])
    if event is None or event[0] != "Result" or len(event[1]) != 2:
        return None
    return {
        "item_type": event[1][0],
        "error_type": event[1][1],
        "lifetime": arguments[0],
    }


def _bytes(source: bytes | str) -> bytes:
    return source if isinstance(source, bytes) else source.encode()


def parse_bindings(types_source: bytes | str, client_source: bytes | str) -> Bindings:
    """Parse generated ``types.rs`` and ``client.rs`` into normalized Bindings."""
    types_source = _bytes(types_source)
    client_source = _bytes(client_source)
    parser = Parser(Language(tree_sitter_rust.language()))

    roots: dict[str, Node] = {}
    symbol_paths: dict[str, str] = {}
    for label, source in (("types", types_source), ("client", client_source)):
        root = parser.parse(source).root_node
        if root.has_error:
            raise ParseError(f"invalid generated Rust {label} syntax")
        roots[label] = root
        for item in root.named_children:
            if item.type not in {"struct_item", "enum_item", "type_item"}:
                continue
            name = _text(source, item.child_by_field_name("name"))
            if name in symbol_paths:
                raise ParseError(f"ambiguous generated symbol {name}")
            symbol_paths[name] = f"crate::generated::{label}::{name}"

    structs: dict[str, list[dict[str, str]]] = {}
    enums: dict[str, list[dict[str, str | None]]] = {}
    aliases: dict[str, str] = {}

    for item in roots["types"].named_children:
        if item.type == "struct_item":
            name = _text(types_source, item.child_by_field_name("name"))
            fields = []
            for field in _children(item.child_by_field_name("body"), "field_declaration"):
                if not _children(field, "visibility_modifier"):
                    continue
                fields.append(
                    {
                        "name": _text(types_source, field.child_by_field_name("name")),
                        "type": _text(types_source, field.child_by_field_name("type")),
                    }
                )
            structs[name] = fields
        elif item.type == "enum_item":
            name = _text(types_source, item.child_by_field_name("name"))
            variants = []
            for variant in _children(item.child_by_field_name("body"), "enum_variant"):
                attributes = [
                    _text(types_source, child)
                    for child in variant.named_children
                    if child.type == "attribute_item"
                ]
                body = variant.child_by_field_name("body")
                variants.append(
                    {
                        "name": _text(types_source, variant.child_by_field_name("name")),
                        "payload": (
                            _text(types_source, body.named_children[0])
                            if body is not None and body.named_children
                            else None
                        ),
                        "wire_name": _serde_rename(attributes),
                    }
                )
            enums[name] = variants
        elif item.type == "type_item":
            name = _text(types_source, item.child_by_field_name("name"))
            aliases[name] = _text(types_source, item.child_by_field_name("type"))

    client_structs = [
        item
        for item in roots["client"].named_children
        if item.type == "struct_item"
    ]
    client_name = (
        _text(client_source, client_structs[0].child_by_field_name("name"))
        if client_structs
        else "ApiClient"
    )
    symbol_paths.setdefault(client_name, f"crate::generated::client::{client_name}")

    operations: dict[str, dict[str, object]] = {}
    for item in roots["client"].named_children:
        if item.type != "impl_item":
            continue
        impl_type = item.child_by_field_name("type")
        if impl_type is None or _text(client_source, impl_type) != client_name:
            continue
        for function in _children(item.child_by_field_name("body"), "function_item"):
            visibility = _children(function, "visibility_modifier")
            if not visibility:
                continue
            name = _text(client_source, function.child_by_field_name("name"))
            parameters = []
            for parameter in _children(function.child_by_field_name("parameters"), "parameter"):
                pattern = parameter.child_by_field_name("pattern")
                type_node = parameter.child_by_field_name("type")
                if pattern is None or type_node is None:
                    continue
                parameter_name = _text(client_source, pattern)
                if parameter_name in {"self", "&self", "&mut self"}:
                    continue
                parameters.append(
                    {
                        "name": parameter_name,
                        "type": _text(client_source, type_node),
                    }
                )
            return_type_node = function.child_by_field_name("return_type")
            return_type = _text(client_source, return_type_node) if return_type_node else "()"
            success_type = return_type
            result = _split_generic(return_type)
            if result is not None and result[0] == "Result" and result[1]:
                success_type = result[1][0]
            operations[name] = {
                "name": name,
                "parameters": parameters,
                "return_type": return_type,
                "success_type": success_type,
                "stream": _stream_binding(success_type),
            }

    value = {
        "schema_version": 2,
        "structs": structs,
        "enums": enums,
        "aliases": aliases,
        "operations": operations,
        "symbol_paths": symbol_paths,
        "binding": {
            "client": {
                "type_path": f"crate::generated::client::{client_name}",
                "constructor": "new",
                "api_key_builder": "with_api_key",
                "base_url_builder": "with_base_url",
            },
            "type_preludes": ["crate::generated::types::*"],
        },
    }
    try:
        return Bindings.from_dict(value)
    except ValueError as error:
        raise ParseError(str(error)) from error


def parse_directory(path: str | Path) -> Bindings:
    """Parse one generated output directory."""
    path = Path(path)
    return parse_bindings(
        (path / "types.rs").read_bytes(),
        (path / "client.rs").read_bytes(),
    )

File: openapi-to-rust-bindings/src/parser.rs
use crate::{Bindings, Error};
use proc_macro2::{LineColumn, Span};
use serde_json::{Map, Value, json};
use std::collections::BTreeMap;
use syn::spanned::Spanned;
use syn::{
    Attribute, Fields, FnArg, GenericArgument, ImplItem, Item, Pat, PathArguments, ReturnType,
    Type, Visibility,
};

struct SourceText<'a> {
    text: &'a str,
    line_starts: Vec<usize>,
}

impl<'a> SourceText<'a> {
    fn new(text: &'a str) -> Self {
        let mut line_starts = vec![0];
        for (index, byte) in text.bytes().enumerate() {
            if byte == b'\n' {
                line_starts.push(index + 1);
            }
        }
        Self { text, line_starts }
    }

    fn slice<T: Spanned + ?Sized>(&self, node: &T) -> Result<String, Error> {
        self.slice_span(node.span())
    }

    fn slice_span(&self, span: Span) -> Result<String, Error> {
        let start = self.offset(span.start())?;
        let end = self.offset(span.end())?;
        self.text
            .get(start..end)
            .map(ToOwned::to_owned)
            .ok_or_else(|| Error::new("missing Rust AST node"))
    }

    fn offset(&self, position: LineColumn) -> Result<usize, Error> {
        let line = position
            .line
            .checked_sub(1)
            .ok_or_else(|| Error::new("missing Rust AST node"))?;
        self.line_starts
            .get(line)
            .and_then(|start| start.checked_add(position.column))
            .filter(|offset| *offset <= self.text.len())
            .ok_or_else(|| Error::new("missing Rust AST node"))
    }
}

fn parse_file(source: &str, label: &str) -> Result<syn::File, Error> {
    syn::parse_file(source)
        .map_err(|_| Error::new(format!("invalid generated Rust {label} syntax")))
}

fn symbol_name(item: &Item) -> Option<String> {
    match item {
        Item::Struct(item) => Some(item.ident.to_string()),
        Item::Enum(item) => Some(item.ident.to_string()),
        Item::Type(item) => Some(item.ident.to_string()),
        _ => None,
    }
}

fn serde_rename(
    attributes: &[Attribute],
    source: &SourceText<'_>,
) -> Result<Option<String>, Error> {
    let mut matches = Vec::new();
    for attribute in attributes {
        let text = source.slice(attribute)?;
        let Some(body) = text
            .strip_prefix("#[serde(")
            .and_then(|value| value.strip_suffix(")]"))
        else {
            continue;
        };
        let Some((name, raw_value)) = body.split_once('=') else {
            continue;
        };
        if name.trim() != "rename" {
            continue;
        }
        let raw_value = raw_value.trim();
        let Some(value) = raw_value
            .strip_prefix('"')
            .and_then(|value| value.strip_suffix('"'))
        else {
            continue;
        };
        if value.contains('"') {
            continue;
        }
        matches.push(value.to_owned());
    }
    if matches.len() > 1 {
        return Err(Error::new(
            "multiple serde rename attributes on enum variant",
        ));
    }
    Ok(matches.pop())
}

fn split_generic(value: &str) -> Option<(String, Vec<String>)> {
    let start = value.find('<')?;
    if !value.ends_with('>') {
        return None;
    }
    let constructor = value[..start].trim().to_owned();
    let body = &value[start + 1..value.len() - 1];
    let mut depth = 0_i32;
    let mut part_start = 0;
    let mut parts = Vec::new();
    for (index, character) in body.char_indices() {
        match character {
            '<' | '(' | '[' | '{' => depth += 1,
            '>' | ')' | ']' | '}' => depth -= 1,
            ',' if depth == 0 => {
                parts.push(body[part_start..index].trim().to_owned());
                part_start = index + character.len_utf8();
            }
            _ => {}
        }
    }
    parts.push(body[part_start..].trim().to_owned());
    Some((constructor, parts))
}

fn stream_binding(success_type: &str) -> Value {
    let Some((constructor, arguments)) = split_generic(success_type) else {
        return Value::Null;
    };
    if constructor != "futures_util::stream::BoxStream" || arguments.len() != 2 {
        return Value::Null;
    }
    let Some((event_constructor, event_arguments)) = split_generic(&arguments[1]) else {
        return Value::Null;
    };
    if event_constructor != "Result" || event_arguments.len() != 2 {
        return Value::Null;
    }
    json!({
        "item_type": event_arguments[0],
        "error_type": event_arguments[1],
        "lifetime": arguments[0],
    })
}

fn success_type(return_type: &Type, source: &SourceText<'_>) -> Result<String, Error> {
    let Type::Path(path) = return_type else {
        return Ok(String::new());
    };
    let Some(segment) = path.path.segments.last() else {
        return Ok(String::new());
    };
    let PathArguments::AngleBracketed(arguments) = &segment.arguments else {
        return Ok(String::new());
    };
    for argument in &arguments.args {
        if let GenericArgument::Type(argument) = argument {
            return source.slice(argument);
        }
    }
    Ok(String::new())
}

fn insert_symbol(
    symbol_paths: &mut BTreeMap<String, String>,
    name: String,
    label: &str,
) -> Result<(), Error> {
    if symbol_paths.contains_key(&name) {
        return Err(Error::new(format!("ambiguous generated symbol {name}")));
    }
    symbol_paths.insert(name.clone(), format!("crate::generated::{label}::{name}"));
    Ok(())
}

/// Parse generated `types.rs` and `client.rs` into normalized Bindings v2.
pub fn parse_bindings(types_source: &str, client_source: &str) -> Result<Bindings, Error> {
    let types_file = parse_file(types_source, "types")?;
    let client_file = parse_file(client_source, "client")?;
    let types_text = SourceText::new(types_source);
    let client_text = SourceText::new(client_source);

    let mut symbol_paths = BTreeMap::new();
    for (label, file) in [("types", &types_file), ("client", &client_file)] {
        for item in &file.items {
            if let Some(name) = symbol_name(item) {
                insert_symbol(&mut symbol_paths, name, label)?;
            }
        }
    }

    let mut structs = Map::new();
    let mut enums = Map::new();
    let mut aliases = Map::new();

    for item in &types_file.items {
        match item {
            Item::Struct(item) => {
                let mut fields = Vec::new();
                if let Fields::Named(named) = &item.fields {
                    for field in &named.named {
                        if !matches!(field.vis, Visibility::Public(_)) {
                            continue;
                        }
                        let name = field
                            .ident
                            .as_ref()
                            .ok_or_else(|| Error::new("missing Rust AST node"))?
                            .to_string();
                        fields.push(json!({
                            "name": name,
                            "type": types_text.slice(&field.ty)?,
                        }));
                    }
                }
                structs.insert(item.ident.to_string(), Value::Array(fields));
            }
            Item::Enum(item) => {
                let mut variants = Vec::new();
                for variant in &item.variants {
                    let payload = match &variant.fields {
                        Fields::Unit => Value::Null,
                        Fields::Unnamed(fields) => {
                            if fields.unnamed.len() != 1 {
                                return Err(Error::new(format!(
                                    "enum {}::{} is not unary",
                                    item.ident, variant.ident
                                )));
                            }
                            Value::String(types_text.slice(&fields.unnamed[0].ty)?)
                        }
                        Fields::Named(_) => Value::Null,
                    };
                    variants.push(json!({
                        "name": variant.ident.to_string(),
                        "payload": payload,
                        "wire_name": serde_rename(&variant.attrs, &types_text)?,
                    }));
                }
                enums.insert(item.ident.to_string(), Value::Array(variants));
            }
            Item::Type(item) => {
                aliases.insert(
                    item.ident.to_string(),
                    Value::String(types_text.slice(&item.ty)?),
                );
            }
            _ => {}
        }
    }

    let mut operations = Map::new();
    let mut found_client = false;
    for item in &client_file.items {
        let Item::Impl(item) = item else {
            continue;
        };
        if client_text.slice(item.self_ty.as_ref())? != "HttpClient" {
            continue;
        }
        found_client = true;
        for item in &item.items {
            let ImplItem::Fn(function) = item else {
                continue;
            };
            let mut parameters = Vec::new();
            for input in &function.sig.inputs {
                let FnArg::Typed(parameter) = input else {
                    continue;
                };
                let Pat::Ident(identifier) = parameter.pat.as_ref() else {
                    return Err(Error::new("missing Rust AST node"));
                };
                parameters.push(json!({
                    "name": identifier.ident.to_string(),
                    "type": client_text.slice(parameter.ty.as_ref())?,
                }));
            }
            let ReturnType::Type(_, return_type) = &function.sig.output else {
                return Err(Error::new("missing Rust AST node"));
            };
            let return_type_text = client_text.slice(return_type.as_ref())?;
            let success_type_text = success_type(return_type.as_ref(), &client_text)?;
            let name = function.sig.ident.to_string();
            operations.insert(
                name.clone(),
                json!({
                    "name": name,
                    "parameters": parameters,
                    "return_type": return_type_text,
                    "success_type": success_type_text,
                    "stream": stream_binding(&success_type_text),
                }),
            );
        }
    }
    if !client_source.trim().is_empty() && !found_client {
        return Err(Error::new("openapi-to-rust HttpClient impl not found"));
    }

    let symbol_paths = symbol_paths
        .into_iter()
        .map(|(name, path)| (name, Value::String(path)))
        .collect::<Map<_, _>>();

    Bindings::from_value(json!({
        "schema_version": 2,
        "structs": structs,
        "enums": enums,
        "aliases": aliases,
        "operations": operations,
        "symbol_paths": symbol_paths,
        "binding": {
            "client": {
                "type_path": "crate::generated::client::HttpClient",
                "constructor": "new",
                "api_key_builder": "with_api_key",
                "base_url_builder": "with_base_url",
            },
            "type_preludes": ["crate::generated::types::*"],
        },
    }))
}
